### PR TITLE
Added Oracle address for each swap pair

### PR DESCRIPTION
**Pull request by glottologist**

### CODE REVIEW

**Comment by glottologist** [comment]
@tj825 this isn't finished but I wanted to get it on your radar early

**Comment by glottologist** [comment]
@tj825 @tomjack For the most part this is done.  I can't really be sure of the scaling of the prices coming from the Oracle as the ghostnet versions aren't producing prices for our pairs yet so I took cues from the mainnet versions of the Harbinger oracle contracts instead.

File: batcher/storage/initial_storage_ghostnet.mligo [hunk 1]
```diff
@@ -23,39 +23,47 @@ let f(_:unit) = {
...
-            standard = Some "FA1.2 token";
-          };
+        swap = {
+            from = {
+              amount = 1n;
+              token = {
+                name = "tzBTC";
+                address = Some(("KT1XLyXAe5FWMHnoWa98xZqgDUyyRms2B3tG" : address));
+                decimals = 8;
+                standard = Some "FA1.2 token";
```
Comment (aguillon) [inline batcher/storage/initial_storage_ghostnet.mligo:33]: I think using something like
`type standard =
  | FA2
  | FA12
  | Other of string`

would be better.

**Review by aguillon — CHANGES_REQUESTED** [review]
Context: I was asked to do this review a bit out of the blue (and don't get me wrong, I liked it), but I am not entirely familiar with last months' progress on Batcher, so I may have missed stuff. In particular I did not run the code, and I am not familiar with the logic for oracles.

Things look OK but the style could be improved a bit.

File: batcher/batcher.mligo [hunk 2]
```diff
@@ -79,6 +79,12 @@ type swap = {
  to : token;
 }
 
+type valid_swap = {
+  swap: swap;
+  oracle: address;
+}
```
[the remark this answers is not in the document]
Comment (glottologist) [reply batcher/batcher.mligo]: Yep each swap is only valid if it has an oracle to source the price from.  

We use the same swap type for swap orders and rates. The amount for orders is whatever is sent by the users and for the exchange rates it is always 1

File: batcher/batcher.mligo [hunk 3]
```diff
@@ -81,7 +81,8 @@ type swap = {
 
 type valid_swap = {
   swap: swap;
-  oracle: address;
+  oracle_address: address;
+  oracle_asset_name: string;
```
[the remark this answers is not in the document]
Comment (glottologist) [reply batcher/batcher.mligo]: The asset names on Harbinger are in the form tzBTC-USDT and we call them using the asset name

[the remark this answers is not in the document]
Comment (glottologist) [reply batcher/batcher.mligo]: There is only one entry getPrice that takes the asset name. They just return a timestamp and nat

File: batcher/batcher.mligo (hunk 2 above)
[the remark this answers is not in the document]
Comment (glottologist) [reply batcher/batcher.mligo]: I added a comment to explain the type

File: batcher/storage/initial_storage_ghostnet.mligo (hunk 1 above)
Comment (glottologist) [reply batcher/storage/initial_storage_ghostnet.mligo:33]: I'll add it as an enhancement issue

Comment (glottologist) [reply batcher/storage/initial_storage_ghostnet.mligo:33]: https://github.com/marigold-dev/batcher/issues/258

File: batcher/batcher.mligo [hunk 4]
```diff
@@ -1367,21 +1362,46 @@ let redeem
...
+                                    let storage = Utils.update_current_rate (rate_name) (oracle_rate) (storage) in
+                                    let pair = Utils.pair_of_rate oracle_rate in
+                                    let current_time = Tezos.get_now () in
+                                    let batch_set = storage.batch_set in
+                                    let (batch_opt, batch_set) = Batch_Utils.get_current_batch_without_opening pair current_time batch_set in
+                                    match batch_opt with
+                                    | Some b -> let batch_set = finalize b current_time oracle_rate batch_set in
+                                               let storage = { storage with batch_set = batch_set } in
+                                               storage
+                                    | None ->   storage)
```
[the remark this answers is not in the document]
Comment (glottologist) [reply batcher/batcher.mligo:1396]: Whilst I agree with this in principle the ligo pretty print isn't a viable solution; it strips out comments which is a maintenance headache.

**Pull request merged**